File: python/ressrf/audit.py
"""Audit event types and sink protocol."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Callable, Protocol, runtime_checkable


@dataclass(frozen=True)
class AuditEvent:
    """Structured audit event emitted by the policy engine.

    Attributes:
        event_type: The kind of event (e.g. 'host_validated', 'url_validated').
        fields: All event fields as a dict.
    """

    event_type: str
    fields: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_json(cls, raw: str) -> AuditEvent:
        """Parse a JSON string into an AuditEvent."""
        data = json.loads(raw)
        event_type = data.pop("event_type", "unknown")
        return cls(event_type=event_type, fields=data)


@runtime_checkable
class AuditSink(Protocol):
    """Protocol for receiving audit events from the policy engine.

    Implement this to route audit events to your logging system.

    Example with logging:

        class LoggingSink:
            def emit(self, event: AuditEvent) -> None:
                logging.info("ressrf.audit", extra={"event": event.fields})

    Example with structlog:

        class StructlogSink:
            def emit(self, event: AuditEvent) -> None:
                structlog.get_logger().info("ressrf.audit", **event.fields)
    """

    def emit(self, event: AuditEvent) -> None: ...


class AuditFunc:
    """Adapts a plain callable to the AuditSink protocol.

    Example:
        sink = AuditFunc(lambda e: print(f"blocked: {e.event_type}"))
    """

    def __init__(self, func: Callable[[AuditEvent], None]) -> None:
        self._func = func

    def emit(self, event: AuditEvent) -> None:
        self._func(event)


class MultiSink:
    """Fans out audit events to multiple sinks."""

    def __init__(self, *sinks: AuditSink) -> None:
        self._sinks = list(sinks)

    def emit(self, event: AuditEvent) -> None:
        for sink in self._sinks:
            sink.emit(event)


class DiscardSink:
    """Silently drops all events."""

    def emit(self, event: AuditEvent) -> None:
        pass


__all__ = ["AuditEvent", "AuditFunc", "AuditSink", "DiscardSink", "MultiSink"]

File: go/ressrf/audit.go
// Package ressrf provides SSRF protection backed by a WASM policy engine.
package ressrf

import (
	"context"
	"encoding/json"
	"time"
)

// AuditEvent represents a structured event emitted by the SSRF policy engine.
type AuditEvent struct {
	Kind      string          `json:"kind"`
	Timestamp time.Time       `json:"timestamp,omitempty"`
	Fields    json.RawMessage `json:"fields,omitempty"`
}

// AuditSink receives audit events from the policy engine.
//
// Implement this interface to route audit events to your logging system.
// For a quick adapter from a plain function, see AuditFunc.
//
// Example with slog:
//
//	sink := ressrf.AuditFunc(func(ctx context.Context, e *ressrf.AuditEvent) {
//	    slog.InfoContext(ctx, "ressrf.audit", "kind", e.Kind, "fields", string(e.Fields))
//	})
//
// Example with zap:
//
//	sink := ressrf.AuditFunc(func(_ context.Context, e *ressrf.AuditEvent) {
//	    logger.Info("ressrf.audit", zap.String("kind", e.Kind))
//	})
type AuditSink interface {
	Emit(ctx context.Context, event *AuditEvent)
}

// AuditFunc adapts a plain function to the AuditSink interface.
type AuditFunc func(ctx context.Context, event *AuditEvent)

// Emit calls the underlying function.
func (f AuditFunc) Emit(ctx context.Context, event *AuditEvent) { f(ctx, event) }

// MultiSink fans out events to multiple sinks.
type MultiSink []AuditSink

// Emit fans out the event to all contained sinks.
func (ms MultiSink) Emit(ctx context.Context, event *AuditEvent) {
	for _, s := range ms {
		s.Emit(ctx, event)
	}
}

// DiscardSink silently drops all events.
type DiscardSink struct{}

// Emit is a no-op.
func (DiscardSink) Emit(context.Context, *AuditEvent) {}
